Explainer for cross-origin trusted signals (#1156)
WICG/turtledove: FLEDGE.md

*   adMetadata: Arbitrary metadata provided by the buyer.
*   bid: A numerical bid value.
*   auctionConfig: The auction configuration object passed to `navigator.runAdAuction()`.
*   trustedScoringSignals: A value retrieved from a real-time trusted server chosen by the seller and reflecting the seller's opinion of this particular creative, as further described in [3.1 Fetching Real-Time Data from a Trusted Server](#31-fetching-real-time-data-from-a-trusted-server) below. This is used when the server is same-origin to the seller; crossOriginTrustedSignals is used otherwise.

> *   Like trustedScoringSignals, but used when the server is cross-origin to the seller script. The

Reviewer: I believe JavaScript doesn't have named function arguments, so I think argument order is quite important.  So can I recommend that we move the description of `crossOriginTrustedSignals` down to after `directFromSellerSignals` so that the description order matches the argument order.

Author: Done

> *   Like trustedScoringSignals, but used when the server is cross-origin to the seller script. The

Author: ... Done. No idea why I didn't have this...

    value is an object that has as a key the trusted server's origin, e.g. `"https://example.org"`,

>     and as value an object in format trustedScoringSignals uses.

Reviewer: ```suggestion
    and as value an object in format `trustedScoringSignals` uses.
```

Author: Done

- [FLEDGE Key/Value Server APIs Explainer](https://github.com/WICG/turtledove/blob/master/FLEDGE_Key_Value_Server_API.md)
- [FLEDGE Key/Value Server Trust Model Explainer](https://github.com/privacysandbox/fledge-docs/blob/main/key_value_service_trust_model.md)

##### 3.1.1 Cross-Origin Trusted Server Signals

If the key-value server is on a different origin than the corresponding script, additional steps are
needed to ensure that sensitive information is not inappropriately leaked to, or is manipulated by,
a third party.

In case of bidder signals, the following changes occur:
1. The fetch is a cross-origin CORS fetch with `Origin:` set to the buyer script's origin.
2. The value is passed to `crossOriginTrustedSignals` parameter, not the `trustedBiddingSignals`

>    parameter, and there is one more level of nesting denoting the server's origin, e.g:

Reviewer 2: question: Is there a benefit to this level of nesting, rather than simply another parameter indicating the origin / domain of the scoring signals?

Author: It's a bit harder to forget to check? (And also less parameters)

Reviewer 3: >harder to forget to check? 

One is already explicitly checking the alternate signals parameter so that should already be priced in, so to speak.

Author: I was thinking in terms of having code do stuff like:
```
trustedSignals = trustedBiddingSignals || crossOriginTrustedSignals["kv-server.mydomain.com"];
```
...which simply can't access  an unexpected domain.

publisher, not the seller:
1. The seller script must provide an `Ad-Auction-Allow-Trusted-Scoring-Signals-From` response header,
   a  [structured headers list of strings](https://www.rfc-editor.org/rfc/rfc8941) describing origins
   fetching trusted signals from which is permitted. The trusted scoring signals fetch may not begin
   until this header is received, in order to avoid leaking bid information to a third party.
   This means using a cross-origin trusted server for seller information may carry a peformance
   penalty.
2. The fetch is a cross-origin CORS fetch with `Origin:` set to the seller script's origin.
3. The value is passed to `crossOriginTrustedSignals` parameter, not the `trustedScoringSignals`
   parameter, and there is one more level of nesting denoting the server's origin, e.g:

```
{
  'https://www.kv-server.example': {
    'renderURL': {'https://cdn.com/render_url_of_bidder': arbitrary_value_from_signals},
    'adComponentRenderURLs': {
        'https://cdn.com/ad_component_of_a_bid': arbitrary_value_from_signals,
        'https://cdn.com/another_ad_component_of_a_bid': arbitrary_value_from_signals,
        ...}
  }
}
```

> Note that older version of Chrome did not support cross-origin trusted signals. You can query

Reviewer 2: question: from a seller's perspective, we won't really have access to an environment where this can be queried before the `auctionConfig` is returned... so how will we know when to leverage this new behaviour?

Author: I guess it would require whatever script that bootstraps you to pass that in?

I wonder if "give me an entire table of settings" alternative to queryFeatureSupport would be helpful here, though I worry it may eventually get rather large.

Reviewer 2: Indeed, this is something that we've been floating on the Prebid.js side for precisely this reason.

Author: So I just landed support for queryFeatureSupport('*') which returns a dictionary with all the features that queryFeatureSupport knows about. Explain update (to the feature detection file) is going to come shortly, just in process of going through the comments here...

Author: ... And pushed.

- [FLEDGE Key/Value Server APIs Explainer](https://github.com/WICG/turtledove/blob/master/FLEDGE_Key_Value_Server_API.md)
- [FLEDGE Key/Value Server Trust Model Explainer](https://github.com/privacysandbox/fledge-docs/blob/main/key_value_service_trust_model.md)

##### 3.1.1 Cross-Origin Trusted Server Signals

If the key-value server is on a different origin than the corresponding script, additional steps are
needed to ensure that sensitive information is not inappropriately leaked to, or is manipulated by,
a third party.

> In case of bidder signals, the following changes occur:

Reviewer: ```suggestion
In the case of bidder signals, the following changes occur:
```

Author: Done

> 1. The fetch is a cross-origin CORS fetch with `Origin:` set to the buyer script's origin.

Reviewer: ```suggestion
1. The fetch is a cross-origin [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) fetch with `Origin:` set to the bidding script's origin.
```

Author: Done

   parameter, and there is one more level of nesting denoting the server's origin, e.g:

```
{
  'https://www.kv-server.example': {
    'keys': {
        'key1': arbitrary_json,
        'key2': arbitrary_json,
        ...},
    'perInterestGroupData': {
        'name1': {
            'priorityVector': {
                'signal1': number,
                'signal2': number,
                ...}
        },
        ...
    }
  }
}
```
3. The data version is passed in `browserSignals.crossOriginDataVersion`, not
   `browserSignals.dataVersion`.

Seller signals have additional requirements, as the `trustedScoringSignalsURL` is provided by the
publisher, not the seller:
1. The seller script must provide an `Ad-Auction-Allow-Trusted-Scoring-Signals-From` response header,
   a  [structured headers list of strings](https://www.rfc-editor.org/rfc/rfc8941) describing origins
   from which fetching trusted signals is permitted. The trusted scoring signals fetch may not begin
   until this header is received, in order to avoid leaking bid information to a third party.
   This means using a cross-origin trusted server for seller information may carry a peformance
   penalty.

> 2. The fetch is a cross-origin CORS fetch with `Origin:` set to the seller script's origin.

Reviewer: ```suggestion
2. The fetch is a cross-origin [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) fetch with `Origin:` set to the seller script's origin.
```

Author: Done

##### 3.1.1 Cross-Origin Trusted Server Signals

If the key-value server is on a different origin than the corresponding script, additional steps are
needed to ensure that sensitive information is not inappropriately leaked to, or is manipulated by,
a third party.

In the case of bidder signals, the following changes occur:
1. The fetch is a cross-origin [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) fetch
   with `Origin:` set to the buyer script's origin.
2. The value is passed to `crossOriginTrustedSignals` parameter, not the `trustedBiddingSignals`
   parameter, and there is one more level of nesting denoting the server's origin, e.g:

```
{
  'https://www.kv-server.example': {
    'keys': {
        'key1': arbitrary_json,
        'key2': arbitrary_json,
        ...},
    'perInterestGroupData': {
        'name1': {
            'priorityVector': {
                'signal1': number,
                'signal2': number,
                ...}
        },
        ...
    }
  }
}
```
3. The data version is passed in `browserSignals.crossOriginDataVersion`, not
   `browserSignals.dataVersion`.

> Seller signals have additional requirements, as the `trustedScoringSignalsURL` is provided by the
> publisher, not the seller:

Reviewer: ```suggestion
Seller signals have additional requirements, as the `trustedScoringSignalsURL` is provided by a
context that is not required to be same-origin with the seller:
```

Author: Done.

   from which fetching trusted signals is permitted. The trusted scoring signals fetch may not begin
   until this header is received, in order to avoid leaking bid information to a third party.
   This means using a cross-origin trusted server for seller information may carry a peformance
   penalty.
2. The fetch is a cross-origin [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) fetch
   with `Origin:` set to the seller script's origin.
3. The value is passed to `crossOriginTrustedSignals` parameter, not the `trustedScoringSignals`
   parameter, and there is one more level of nesting denoting the server's origin, e.g:

```
{
  'https://www.kv-server.example': {
    'renderURL': {'https://cdn.com/render_url_of_bidder': arbitrary_value_from_signals},
    'adComponentRenderURLs': {
        'https://cdn.com/ad_component_of_a_bid': arbitrary_value_from_signals,
        'https://cdn.com/another_ad_component_of_a_bid': arbitrary_value_from_signals,
        ...}
  }
}
```
4. The data version is passed in `browserSignals.crossOriginDataVersion`, not
   `browserSignals.dataVersion`.

> Note that older version of Chrome did not support cross-origin trusted signals. You can query

Reviewer: ```suggestion
Note that older versions of Chrome did not support cross-origin trusted signals. You can query
```

Author: Done

##### 3.1.1 Cross-Origin Trusted Server Signals

If the key-value server is on a different origin than the corresponding script, additional steps are
needed to ensure that sensitive information is not inappropriately leaked to, or is manipulated by,
a third party.

In case of bidder signals, the following changes occur:
1. The fetch is a cross-origin CORS fetch with `Origin:` set to the buyer script's origin.
2. The value is passed to `crossOriginTrustedSignals` parameter, not the `trustedBiddingSignals`
   parameter, and there is one more level of nesting denoting the server's origin, e.g:

```
{
  'https://www.kv-server.example': {
    'keys': {
        'key1': arbitrary_json,
        'key2': arbitrary_json,
        ...},
    'perInterestGroupData': {
        'name1': {
            'priorityVector': {
                'signal1': number,
                'signal2': number,
                ...}
        },
        ...
    }
  }
}
```

> Seller signals have additional requirements, as the `trustedScoringSignalsURL` is provided by the

Reviewer 3: This is true in the case where the publisher is invoking runAdAuction and is the top-level seller. But otherwise, sellers are providing & operating the TSS url.

Author: It's still in context of the publisher page, and thus not really having the full trust of the seller's origin though?

publisher, not the seller:
1. The seller script must provide an `Ad-Auction-Allow-Trusted-Scoring-Signals-From` response header,
   a  [structured headers list of strings](https://www.rfc-editor.org/rfc/rfc8941) describing origins

>    fetching trusted signals from which is permitted. The trusted scoring signals fetch may not begin

Reviewer 3: describing origins from which fetching trusted signals is permitted

Author: Applied, thanks

   until this header is received, in order to avoid leaking bid information to a third party.

>    This means using a cross-origin trusted server for seller information may carry a peformance

Reviewer 3: >This means using a cross-origin trusted server for seller information may carry a peformance penalty.

I anticipate that a common seller use case will be a TSS URL that is _same-site_ with the scoring logic URL. Must this case incur the penalty?

Reviewer 2: question: given that the original cross-origin fetch is CORS-based, how does `Access-Control-Max-Age` come into play here?  In other words, can we provide a mechanism to cache this granting of cross-site access?  As @dmdabbs correctly noted, the use case envisioned is a seller using different sub-domains from the same eTLD+1, and to avoid additional performance penalties as a result of this.

Reviewer 2: Or, better yet, for those "cross-origin" sharing eLTD+1, is there a way to avoid this penalty entirely?

Author: Hmm, I need to figure out who to discuss this with. Your concern is eminently reasonable, but also generally it feels like web platform has been trying to be more origin than eTLD+1 based.

I think Access-Control-Max-Age doesn't come in at all since the signals fetch doesn't need a preflight, though I guess you're suggesting that the script headers sort of play the same role?

Reviewer 2: I'm trying to better understand if we have pay this penlaty on each auction -- my comment about CORS was simply that we're borrowing those ACAO concepts.

Reviewer 2: On today's WICG meeting, it was mentioned that perhaps the pre-connect may still begin, but it's the fetch() itself that is gated -- can we clarify?

Author: Preconnect should still work, only the load is gated, yes. (I should probably test that, though; this is based on staring at the code really hard).

*   perBuyerSignals: The value for _this specific buyer_ as taken from the auction config passed to `runAdAuction()`.  This can include contextual signals about the page that come from the buyer's server, if the seller is an SSP which performs a real-time bidding call to buyer servers and pipes the response back, or if the publisher page contacts the buyer's server directly.  If so, the buyer may wish to check a cryptographic signature of those signals inside `generateBid()` as protection against tampering.
*   trustedBiddingSignals: An object whose keys are the `trustedBiddingSignalsKeys` for the interest group, and whose values are those returned in the `trustedBiddingSignals` request. This used when the trusted server is same-origin with the buyer's script.
*   crossOriginTrustedSignals: Like trustedBiddingSignals, but used when the trusted-server is
    cross-origin to the buyer's script. The value is an object that has as a key the trusted

>     server's origin, e.g. `"https://example.org"`, and as value an object in format trustedBiddingSignals uses.

Reviewer 3: >...a key the trusted server's origin, e.g. `"https://example.org"`

Suggest using the same example value as in the preceding markup for consistency as well as it is clearly an origin.

Author: Done

*   auctionSignals: As provided by the seller in the call to `runAdAuction()`.  This is the opportunity for the seller to provide information about the page context (ad size, publisher ID, etc), the type of auction (first-price vs second-price), and so on.
*   perBuyerSignals: The value for _this specific buyer_ as taken from the auction config passed to `runAdAuction()`.  This can include contextual signals about the page that come from the buyer's server, if the seller is an SSP which performs a real-time bidding call to buyer servers and pipes the response back, or if the publisher page contacts the buyer's server directly.  If so, the buyer may wish to check a cryptographic signature of those signals inside `generateBid()` as protection against tampering.
*   trustedBiddingSignals: An object whose keys are the `trustedBiddingSignalsKeys` for the interest group, and whose values are those returned in the `trustedBiddingSignals` request. This used when the trusted server is same-origin with the buyer's script.

> *   crossOriginTrustedSignals: Like trustedBiddingSignals, but used when the trusted-server is

Reviewer: I believe JavaScript doesn't have named function arguments, so I think argument order is quite important.  So can I recommend that we move the description of `crossOriginTrustedSignals` down to after `directFromSellerSignals` so that the description order matches the argument order.

Author: Done

> *   crossOriginTrustedSignals: Like trustedBiddingSignals, but used when the trusted-server is

Reviewer: ```suggestion
*   crossOriginTrustedSignals: Like `trustedBiddingSignals`, but used when the trusted-server is
```

Author: Done

    cross-origin to the buyer's script. The value is an object that has as a key the trusted

>     server's origin, e.g. `"https://www.kv-server.example"`, and as value an object in format trustedBiddingSignals uses.

Reviewer: ```suggestion
    server's origin, e.g. `"https://www.kv-server.example"`, and as value an object in format `trustedBiddingSignals` uses.
```

Author: Done
